libsel4/include/sel4/bootinfo_types.h

```diff
@@ -69,9 +69,7 @@ typedef struct seL4_BootInfo {
     seL4_SlotRegion   schedcontrol; /* Caps to sched_control for each node */
 #endif
     seL4_SlotRegion   untyped;         /* untyped-object caps (untyped caps) */
-    seL4_UntypedDesc  untypedList[CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS]; /* information about each untyped */
-    /* the untypedList should be the last entry in this struct, in order
-     * to make this struct easier to represent in other languages */
+    seL4_UntypedDesc  untypedList[];   /* information about each untyped */
 } seL4_BootInfo;
 
```

Comment:
Wouldn't it make more sense to define MAX_NUM_BOOTINFO_UNTYPED_CAPS in the kernel underneath the seL4_BootInfo structure definition and define CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS  to that?

Reply:
I wanted to do that, but have not found a way yet how to do it. It's a bit of a chicken-egg problem to calculate this. And it the end I wonder if all this is really worth the effort and who really needs `CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS`. In the boot info's `seL4_SlotRegion untyped` you get the start and end slot, calculating the amount of descriptors from the delta is easy. The kernel will also guarantee that the start and end values in there is consistent with the items in the array. That just leave the corner case that there are more untyped to be created than there is space for descriptors in the boot info block, which is to be discussed in https://github.com/seL4/seL4/issues/716

Reply:
Inline would be best but also the ugliest, but you could just add a:
```
#define MAX_NUM_BOOTINFO_UNTYPED_CAPS ((SEL4_BIT(BI_FRAME_SIZE_BITS) - sizeof(struct seL4_BootInfo)) / sizeof(seL4_UntypedDesc))
```
underneath it, no need for offsetof() either then.

Reply:
Or perhaps have no CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS or any define at all and just use sizeof() in the one place that uses it instead of offsetof().

Reply:
> Inline would be best but also the ugliest, but you could just add a:
> ```
> #define MAX_NUM_BOOTINFO_UNTYPED_CAPS ((SEL4_BIT(BI_FRAME_SIZE_BITS) - sizeof(struct seL4_BootInfo)) / sizeof(seL4_UntypedDesc))
> ```
> underneath it, no need for offsetof() either then.

I'm not sure if this really works. I have though of splitting up `seL4_BootInfo` and declare the fixed part separately. Then it is known and a macro can be used to calculate the elements for an additional variable part.
```
typedef struct seL4_BootInfo_flex {
    seL4_Word         extraLen;
    ...
    seL4_SlotRegion untyped;
} seL4_BootInfo_fixed

#define CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS ((SEL4_BIT(BI_FRAME_SIZE_BITS) - sizeof(seL4_BootInfo_fixed)) / sizeof(seL4_UntypedDesc))

typedef struct seL4_BootInfo {
    // ... and here I'm lost how to do the magic to get seL4_BootInfo_fixed in here without 
    // giving it a name. We could make the struct boy a big macro that is used twice....

    seL4_UntypedDesc  untypedList[CONFIG_MAX_NUM_BOOTINFO_UNTYPED_CAPS];

} seL4_BootInfo_fixed

Reply:
Just leave it as ``seL4_UntypedDesc  untypedList[];`` and it works fine?

Reply:
Ah, should have read this discussion first before commenting above. You can't have a non-fixed struct size anyway, you'll need to compute a size here for declaration. That is the reason it was a config parameter in the first place, because it's annoying to do that.